OptimoveSDK/Sources/Classes/Optimobile/AnalyticsHelper.swift

```diff
@@ -13,34 +13,41 @@ class KSEventModel: NSManagedObject {
 
 typealias SyncCompletedBlock = (Error?) -> Void
 
-class AnalyticsHelper {
+final class AnalyticsHelper {
+    let eventsHttpClient: KSHttpClient
     private var analyticsContext: NSManagedObjectContext?
     private var migrationAnalyticsContext: NSManagedObjectContext?
-    private var eventsHttpClient: KSHttpClient
+    private var finishedInitializationToken: NSObjectProtocol?
 
     // MARK: Initialization
 
-    init(apiKey: String, secretKey: String, baseEventsUrl: String) {
+    init(httpClient: KSHttpClient) {
         analyticsContext = nil
         migrationAnalyticsContext = nil
 
-        eventsHttpClient = KSHttpClient(baseUrl: URL(string: baseEventsUrl)!, requestFormat: .json, responseFormat: .json)
-        eventsHttpClient.setBasicAuth(user: apiKey, password: secretKey)
+        eventsHttpClient = httpClient
 
         initContext()
 
-        DispatchQueue.global().async {
-            if self.migrationAnalyticsContext != nil {
-                self.syncEvents(context: self.migrationAnalyticsContext)
+        finishedInitializationToken = NotificationCenter.default
+            .addObserver(forName: .optimobileInializationFinished, object: nil, queue: nil) { [weak self] notification in
+                guard let self = self else { return }
+                self.flushEvents()
```

Comment:
Does it matter that we're changing thread semantics here?

Reply:
It does not matter because the actual work will be done on CoreData's context thread.

```swift 
    private func syncEvents(context: NSManagedObjectContext?, _ onSyncComplete: SyncCompletedBlock? = nil) {
        context?.performAndWait {
            let results = fetchEventsBatch(context)
```

https://github.com/optimove-tech/Optimove-SDK-iOS/blob/8d6bccac9cbbf6ce76da4a857d0f84554dcadf1d/OptimoveSDK/Sources/Classes/Optimobile/AnalyticsHelper.swift#L163C18-L163C18